documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+##### Example
+In `array + foo(x)` where `foo(x)` is scalar, foo will only be called once.
+##### Plans to enforce the rule while lowering to FIR
+The evaluation of scalar operands will be inserted before the loop evaluating the elemental expression. In the example above, the generated FIR should look like:
+```
+%1 = fir.call foo %x
+fir.loop %i 1 to n {
+  %2 = fir.extract_value %array %i 
+  %3 = fir.fadd  %2 %1
+  %4 = fir.coordinate_of %result %i
+  fir.store %3 to %4
+} 
+```
+#### Short-circuiting
+##### Rule
+Expression lowering to FIR must insert short-circuiting of operand evaluation that are guaranteed to not be performed in contexts defined in by f18 semantic extensions. Other short-circuiting opportunities should be left to later optimization phases.
+##### Rational
+If it needs to be guaranteed something will not be evaluated if unneeded, then that is semantic related, and it is unsafe to assume later FIR transformation and processing will respect this if short-circuiting was not explicitly inserted while lowering the expression.
+However, in general, the decision of whether to short-circuit an evaluation or not should be left to the optimizers based on the cost of the useless operation compared to the cost of the potential branch. Therefore, expression lowering should not add short-circuits that are not required because these cannot always be reversed in FIR (unless FIR transformations take the liberty to violate semantics which they probably should not).
+Note that it is unclear how well FIR transformation will be able to introduce short-circuits that are rather obvious looking at the `evaluate::Expr` tree, so this rule might get revisited.
+So far, the required short-circuits are:
+- `b.OR.SE`, `b.AND.SE` -> do not evaluate `SE` when `b` is false and `SE` has side effects.
+- `MERGE(SE1, SE2, b)` -> when `b` is a scalar, either `SE1` or `SE2` evaluation has side effects.
+Side effects should be understood in the broad sense (array accesses can cause segfaults, floating point operations can raise signals).
+##### Example
+In `b.AND.foo()` the call to `foo` shall not be performed if `b` is false.
+##### Plans to enforce the rule while lowering to FIR
+It is possible to use `fir.where` to implement a simple if/else logic without having to build a control flow graph using `br_cond` and basic blocks. Although both of these implementation are equivalent, `fir.where` gives more structural guarantees than basic block logic and is expected to interact in better way with optimizations. It is also more readable.
+However, with a `fir.where`, the result of such expression will need to be materialized (with basic blocks logic, one could have taken advantages of phi nodes scalar values, but FIR/LLVM transformation should anyway get rid of scalar temporary in such cases).
```

Comment:
PHI nodes are related to phi functions in IR with static single assignment form property. Phi functions are used to decide what value an SSA node has in a basic block depending on which other basic block it was entered from. This is to implement something like `if (b){i=1} else {i=2}; j=i;)` without using memory operation for the variable `i`.
Now, it is good that you raise this because LLVM uses phi nodes (https://github.com/tensorflow/mlir/blob/master/g3doc/LangRef.md#blocks) but MLIR uses block arguments (https://github.com/tensorflow/mlir/blob/master/g3doc/LangRef.md#blocks) which are IMHO easier to understand.

So I have rewritten this to:
"one could have taken advantages of [MLIR basic block arguments](https://github.com/tensorflow/mlir/blob/master/g3doc/LangRef.md#blocks) to avoid materialization"

FYI, the illustration of `if (b){i=1} else {i=2}; j=i;)` with basic block arguments where we do need to materialize `i` in memory could be:
```
    cond_br %b ^bb1(1) ^bb2(2)
^bb1(%i : i32)  // %i is a block argument, and we would need a PHI node for it in LLVM IR.
    fir.store %i to %j
```